optuna/storages/_heartbeat.py

```diff
@@ -68,6 +72,76 @@ def get_failed_trial_callback(self) -> Optional[Callable[["optuna.Study", Frozen
         raise NotImplementedError()
 
 
+class BaseHeartbeatThread(metaclass=abc.ABCMeta):
+    def __enter__(self) -> None:
+        self.start()
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[Exception]],
+        exc_value: Optional[Exception],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        self.join()
+
+    @abc.abstractmethod
+    def start(self) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def join(self) -> None:
+        raise NotImplementedError()
+
+
+class NullHeartbeatThread(BaseHeartbeatThread):
+    def __init__(self) -> None:
+        pass
+
+    def start(self) -> None:
+        pass
+
+    def join(self) -> None:
+        pass
+
+
+class HeartbeatThread(BaseHeartbeatThread):
+    def __init__(self, trial_id: int, heartbeat: BaseHeartbeat) -> None:
+        self._trial_id = trial_id
+        self._heartbeat = heartbeat
+        self._thread: Optional[Thread] = None
+        self._stop_event: Optional[Event] = None
+
+    def start(self) -> None:
+        self._stop_event = Event()
+        self._thread = Thread(
+            target=self._record_heartbeat, args=(self._trial_id, self._heartbeat, self._stop_event)
+        )
+        self._thread.start()
+
+    def join(self) -> None:
+        assert self._stop_event is not None
+        assert self._thread is not None
+        self._stop_event.set()
+        self._thread.join()
```

Comment:
After offline discussion with @toshihikoyanase and @contramundum53, we couldn't fully reach consensus: `stop`, `join`, `stop_and_join`, `exit`.

Since this choice is mostly a matter of taste, let's revisit it later, when we expose this API to the user. So far, I think we can just leave it `join`.